File: src/main/java/com/silverboxsoft/dynamodbtool/controller/inputdialog/AbsDynamoDbInputDialog.java
package com.silverboxsoft.dynamodbtool.controller.inputdialog;

import java.util.ArrayList;
import java.util.List;

import javafx.beans.value.ChangeListener;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.geometry.Rectangle2D;
import javafx.scene.Node;
import javafx.scene.control.Button;
import javafx.scene.control.ButtonBar.ButtonData;
import javafx.scene.control.ButtonType;
import javafx.scene.control.Dialog;
import javafx.scene.control.DialogEvent;
import javafx.scene.control.Label;
import javafx.scene.control.OverrunStyle;
import javafx.scene.control.RadioButton;
import javafx.scene.control.ScrollPane;
import javafx.scene.control.ScrollPane.ScrollBarPolicy;
import javafx.scene.control.ToggleGroup;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.ColumnConstraints;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.RowConstraints;
import javafx.stage.Screen;
import javafx.stage.Stage;
import javafx.stage.Window;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;

public abstract class AbsDynamoDbInputDialog<R> extends Dialog<R> {

	protected static final String VALIDATION_MSG_INVALID_VALUE = "Invalid value.";

	private final GridPane headGridPane;
	private final GridPane gridPane;
	private final GridPane footGridPane;
	private final ScrollPane scrollPane;
	private final AnchorPane gridAnchorPane;
	private final Button addButton;

	private ButtonData buttonData;
	private boolean isValidData = false;

	private R dynamoDbRecordOrg;
	List<List<Node>> orgBodyAttribueNodeList = new ArrayList<>();
	List<List<Node>> addBodyAttribueNodeList = new ArrayList<>();

	protected static final int HGAP = 5;
	protected static final int VGAP = 5;
	protected static final int FILELD_WIDTH = 300;
	protected static final int NAME_COL_WIDTH = 150;
	protected static final int DEL_COL_WIDTH = 50;
	protected static final int GRID_MIN_HEIGHT = 25;

	protected static final String STRFLD_ID_PREFIX = "txtEdit_";
	protected static final String NUMFLD_ID_PREFIX = "numEdit_";
	protected static final String BINFLD_ID_PREFIX = "binEdit_";
	protected static final String VALLBL_ID_PREFIX = "valLabel_";
	protected static final String EDTBTN_ID_PREFIX = "btnEdit_";
	protected static final String ADDBTN_ID = "btnAddAttribute";
	protected static final String DEL_ID_PREFIX = "ckbDel_";
	protected static final String BTN_TITLE = "Edit";

	public AbsDynamoDbInputDialog(R dynamoDbRecord, String dialogTitle) {
		this.setResizable(true);
		this.getDialogPane().getButtonTypes().addAll(ButtonType.OK, ButtonType.CANCEL);
		this.setTitle(dialogTitle);
		this.setOnCloseRequest(this::doFinalConfirmation);

		dynamoDbRecordOrg = dynamoDbRecord;

		setResultConverter((dialogButton) -> {
			buttonData = dialogButton == null ? null : dialogButton.getButtonData();
			if (buttonData == ButtonData.OK_DONE) {
				isValidData = isFinalValidationOk();
				return isValidData ? getEditedDynamoDbRecord() : getEmptyAttr();
			}
			return null;
		});

		headGridPane = new GridPane();
		gridPane = new GridPane();
		gridAnchorPane = new AnchorPane();
		scrollPane = new ScrollPane();
		footGridPane = new GridPane();
		addButton = new Button("Add");
		setupComponent();

		initialize();
		setColumnConstraints();
		setupScreen();
	}

	/*
	 * for setup
	 */
	abstract List<Integer> getHeaderWidthList();

	abstract List<Node> getHeaderLabelList();

	abstract List<List<Node>> getBodyAttributeNodeList();

	abstract List<Node> getFooterNodeList();

	/*
	 * for screen resize
	 */
	abstract int getValueColIndex();

	/*
	 * for return dialog result
	 */
	abstract R getEditedDynamoDbRecord();

	abstract R getEmptyAttr();

	/*
	 * for add attribute action
	 */
	abstract void actAddNewAttribute();

	/*
	 * for final validation
	 */
	abstract boolean isFinalValidationOk();

	/*
	 * accessor for inherited class
	 */
	protected GridPane getGridPane() {
		return this.gridPane;
	}

	protected R getDynamoDbRecordOrg() {
		return dynamoDbRecordOrg;
	}

	protected GridPane getFootGridPane() {
		return footGridPane;
	}

	protected Button getAddButton() {
		return addButton;
	}

	protected ButtonData getButtonData() {
		return buttonData;
	}

	/*
	 * utility function
	 */
	protected List<List<Node>> getCurrentBodyNodeList() {
		List<List<Node>> curNodeList = new ArrayList<>();
		curNodeList.addAll(orgBodyAttribueNodeList);
		curNodeList.addAll(addBodyAttribueNodeList);
		return curNodeList;
	}

	protected Label getContentLabel(String text, boolean isBold) {
		Label label = getContentLabel(text);
		label.setStyle("-fx-font-weight: bold;");
		return label;
	}

	/*
	 * static function
	 */
	// same as DialogPane.createContentLabel
	protected static Label getContentLabel(String text) {
		Label label = new Label(text);
		label.setMaxWidth(Double.MAX_VALUE);
		label.setMaxHeight(Double.MAX_VALUE);
		label.getStyleClass().add("content");
		label.setWrapText(false);
		label.setTextOverrun(OverrunStyle.ELLIPSIS);
		label.setText(text);
		return label;
	}

	protected static HBox getBooleanInput(AttributeValue attrVal) {
		HBox hbox = new HBox(HGAP);
		RadioButton radioButtonTrue = new RadioButton();
		RadioButton radioButtonFalse = new RadioButton();
		radioButtonTrue.setText("TRUE");
		radioButtonFalse.setText("FALSE");
		ToggleGroup toggleGroup = new ToggleGroup();
		radioButtonTrue.setToggleGroup(toggleGroup);
		radioButtonFalse.setToggleGroup(toggleGroup);
		radioButtonTrue.setSelected(attrVal.bool());
		radioButtonFalse.setSelected(!attrVal.bool());
		hbox.getChildren().addAll(radioButtonTrue, radioButtonFalse);
		return hbox;
	}

	protected static Node getNullViewLabel() {
		Label vallabel = getContentLabel("<null>");
		return vallabel;
	}

	protected static boolean getBooleanValue(HBox hbox) {
		RadioButton radioButtonTrue = (RadioButton) hbox.getChildren().get(0);
		return radioButtonTrue.isSelected();
	}

	protected static Node addUnderlineStyleToNode(Node node) {
		String wkStyle = node.getStyle();
		StringBuilder sb_style = new StringBuilder(wkStyle);
		sb_style.append("-fx-border-style: none none solid none;");
		sb_style.append("-fx-border-color: white white lightgray white;");
		sb_style.append("-fx-border-width: 0 0 1 0;");
		node.setStyle(sb_style.toString());
		return node;
	}

	protected void doFinalConfirmation(DialogEvent dialogEvent) {
		if (buttonData != ButtonData.OK_DONE) {
			return;
		}
		if (!isValidData) {
			dialogEvent.consume();
		}
	}

	/*
	 * class methods
	 */
	protected void setupComponent() {

		headGridPane.setHgap(HGAP);
		headGridPane.setVgap(VGAP);
		headGridPane.setMaxWidth(Double.MAX_VALUE);
		headGridPane.setMinHeight(GRID_MIN_HEIGHT);
		headGridPane.setAlignment(Pos.CENTER);

		gridPane.setHgap(HGAP);
		gridPane.setVgap(VGAP);
		gridPane.setMaxWidth(Double.MAX_VALUE);
		gridPane.setAlignment(Pos.CENTER);
		gridPane.setMinHeight(GRID_MIN_HEIGHT);
		// gridPane.setStyle();
		gridAnchorPane.getChildren().add(gridPane);
		AnchorPane.setLeftAnchor(gridPane, 0.);
		AnchorPane.setTopAnchor(gridPane, 5.);
		AnchorPane.setRightAnchor(gridPane, 0.);
		AnchorPane.setBottomAnchor(gridPane, 5.);
		scrollPane.setContent(gridAnchorPane);

		footGridPane.setHgap(HGAP);
		footGridPane.setVgap(VGAP);
		footGridPane.setMaxWidth(Double.MAX_VALUE);
		footGridPane.setMinHeight(GRID_MIN_HEIGHT);
		footGridPane.setAlignment(Pos.CENTER);
		// footGridPane.setStyle("-fx-padding-top: 15;-fx-spacing-top: 10;-fx-background-color:green");
		HBox footBox = new HBox();
		Insets footInset = new Insets(5., 0., 0., 0.);
		HBox.setMargin(footGridPane, footInset);
		footBox.getChildren().add(footGridPane);

		BorderPane borderPane = new BorderPane();
		borderPane.setCenter(scrollPane);
		borderPane.setTop(headGridPane);
		borderPane.setBottom(footBox);

		addButton.setId(ADDBTN_ID);
		addButton.setOnAction((event) -> {
			actAddNewAttribute();
		});
		getDialogPane().setContent(borderPane);
	}

	protected void initialize() {

		headGridPane.getChildren().clear();
		List<Node> labelList = getHeaderLabelList();
		for (int cIdx = 0; cIdx < labelList.size(); cIdx++) {
			headGridPane.add(labelList.get(cIdx), cIdx, 0);
		}

		getGridPane().getChildren().clear();
		orgBodyAttribueNodeList = getBodyAttributeNodeList();
		for (int rIdx = 0; rIdx < orgBodyAttribueNodeList.size(); rIdx++) {
			List<Node> nodelList = orgBodyAttribueNodeList.get(rIdx);
			for (int cIdx = 0; cIdx < nodelList.size(); cIdx++) {
				getGridPane().add(nodelList.get(cIdx), cIdx, rIdx);
			}
			RowConstraints wkRowConstraint = new RowConstraints();
			wkRowConstraint.setMinHeight(GRID_MIN_HEIGHT);
			getGridPane().getRowConstraints().add(wkRowConstraint);
		}

		updateFooter();
	}

	protected void setupScreen() {
		Rectangle2D screenSize2d = Screen.getPrimary().getVisualBounds();
		double screenHeight = screenSize2d.getHeight();
		double screenWidth = screenSize2d.getWidth();

		final Window window = this.getDialogPane().getScene().getWindow();
		Stage stage = (Stage) window;
		stage.setMinHeight(200);
		stage.setMinWidth(300);
		stage.setMaxHeight(screenHeight);
		stage.setMaxWidth(screenWidth);
	}

	private void setColumnConstraints() {
		int offset = 0;
		for (int idx = 0; idx < getHeaderWidthList().size(); idx++) {
			Integer wkWidth = getHeaderWidthList().get(idx);
			ColumnConstraints wkColConstraint = new ColumnConstraints();
			wkColConstraint.setPrefWidth(wkWidth);
			gridPane.getColumnConstraints().add(wkColConstraint);
			headGridPane.getColumnConstraints().add(wkColConstraint);
			footGridPane.getColumnConstraints().add(wkColConstraint);
			if (idx != getValueColIndex()) {
				offset += wkWidth;
			}
		}

		scrollPane.setHbarPolicy(ScrollBarPolicy.NEVER);
		final int finalOffset = offset;
		ChangeListener<Number> stageSizeListener = (observable, oldValue, newValue) -> {
			double newColWidth = newValue.doubleValue() - finalOffset - HGAP * (getHeaderWidthList().size() - 1);
			gridPane.getColumnConstraints().get(getValueColIndex()).setPrefWidth(newColWidth);
			headGridPane.getColumnConstraints().get(getValueColIndex()).setPrefWidth(newColWidth);
			footGridPane.getColumnConstraints().get(getValueColIndex()).setPrefWidth(newColWidth);
		};
		scrollPane.widthProperty().addListener(stageSizeListener);
	};

	protected void addAttributeNodeList(List<Node> newNodeList) {
		int newIdx = orgBodyAttribueNodeList.size() + addBodyAttribueNodeList.size();
		for (int cIdx = 0; cIdx < newNodeList.size(); cIdx++) {
			getGridPane().add(newNodeList.get(cIdx), cIdx, newIdx);
		}
		addBodyAttribueNodeList.add(newNodeList);
	};

	protected void updateFooter() {
		footGridPane.getChildren().clear();
		List<Node> footerNode = getFooterNodeList();
		for (int cIdx = 0; cIdx < footerNode.size(); cIdx++) {
			footGridPane.add(footerNode.get(cIdx), cIdx, 0);
		}
	}
}

File: src/main/java/com/silverboxsoft/dynamodbtool/controller/inputdialog/AbsDynamoDbInputDialog.kt
package com.silverboxsoft.dynamodbtool.controller.inputdialog

import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.control.ButtonBar.ButtonData
import javafx.scene.control.ScrollPane.ScrollBarPolicy
import javafx.scene.layout.*
import javafx.stage.Screen
import javafx.stage.Stage
import software.amazon.awssdk.services.dynamodb.model.AttributeValue

abstract class AbsDynamoDbInputDialog<R>(dynamoDbRecord: R, dialogTitle: String) : Dialog<R>() {
    private val headGridPane: GridPane

    /*
	 * accessor for inherited class
	 */
    private val gridPane: GridPane
    private val footGridPane: GridPane
    private val scrollPane: ScrollPane
    private val gridAnchorPane: AnchorPane
    protected val addButton: Button
    protected var buttonData: ButtonData = ButtonData.OTHER
    private var isValidData = false
    protected val dynamoDbRecordOrg: R
    var orgBodyAttributeNodeList: List<List<Node>> = ArrayList()
    var addBodyAttributeNodeList: MutableList<List<Node>> = ArrayList()

    /*
	 * for setup
	 */
    abstract val headerWidthList: List<Int>
    abstract val headerLabelList: List<Node>
    abstract val bodyAttributeNodeList: List<List<Node>>

    /*
	 * for screen resize
	 */
    abstract val valueColIndex: Int

    /*
	 * for return dialog result
	 */
    abstract val editedDynamoDbRecord: R
    abstract val emptyAttr: R
    abstract fun isValueChanged(): Boolean
    abstract fun isAddValueRemain(): Boolean

    /*
	 * for add attribute action
	 */
    abstract fun getFooterNodeList(): List<Node>
    abstract fun actAddNewAttribute()

    /*
	 * for final validation
	 */
    abstract val isFinalValidationOk: Boolean

    /*
	 * utility function
	 */
    protected val currentBodyNodeList: List<List<Node>>
        get() {
            val curNodeList: MutableList<List<Node>> = ArrayList()
            curNodeList.addAll(orgBodyAttributeNodeList)
            curNodeList.addAll(addBodyAttributeNodeList)
            return curNodeList
        }

    protected fun getContentLabel(text: String, isBold: Boolean): Label {
        val label = getContentLabel(text)
        label.style = "-fx-font-weight: bold;"
        return label
    }

    protected open fun doFinalConfirmation(dialogEvent: DialogEvent) {
        if (buttonData == ButtonData.OK_DONE) {
            if (isAddValueRemain()){
                val dialog = Dialog<ButtonType>()
                dialog.contentText = "It's seems forgot to add the value. it's OK to close?"
                dialog.dialogPane.buttonTypes.addAll(ButtonType.YES, ButtonType.NO)
                dialog.showAndWait().ifPresent { buttonType: ButtonType ->
                    if (buttonType != ButtonType.YES) {
                        dialogEvent.consume()
                    }
                }
            }
        } else {
            if (isValueChanged() || isAddValueRemain()){
                val dialog = Dialog<ButtonType>()
                dialog.contentText = "Some value is changed. it's OK to cancel?"
                dialog.dialogPane.buttonTypes.addAll(ButtonType.YES, ButtonType.NO)
                dialog.showAndWait().ifPresent { buttonType: ButtonType ->
                    if (buttonType != ButtonType.YES) {
                        dialogEvent.consume()
                    }
                }
            }
            return // close
        }
        if (!isValidData) {
            dialogEvent.consume() // cancel
        }
    }

    /*
	 * class methods
	 */
    private fun setupComponent() {
        headGridPane.hgap = HGAP.toDouble()
        headGridPane.vgap = VGAP.toDouble()
        headGridPane.maxWidth = Double.MAX_VALUE
        headGridPane.minHeight = GRID_MIN_HEIGHT.toDouble()
        headGridPane.alignment = Pos.CENTER
        gridPane.hgap = HGAP.toDouble()
        gridPane.vgap = VGAP.toDouble()
        gridPane.maxWidth = Double.MAX_VALUE
        gridPane.alignment = Pos.CENTER
        gridPane.minHeight = GRID_MIN_HEIGHT.toDouble()
        // gridPane.setStyle();
        gridAnchorPane.children.add(gridPane)
        AnchorPane.setLeftAnchor(gridPane, 0.0)
        AnchorPane.setTopAnchor(gridPane, 5.0)
        AnchorPane.setRightAnchor(gridPane, 0.0)
        AnchorPane.setBottomAnchor(gridPane, 5.0)
        scrollPane.content = gridAnchorPane
        footGridPane.hgap = HGAP.toDouble()
        footGridPane.vgap = VGAP.toDouble()
        footGridPane.maxWidth = Double.MAX_VALUE
        footGridPane.minHeight = GRID_MIN_HEIGHT.toDouble()
        footGridPane.alignment = Pos.CENTER
        // footGridPane.setStyle("-fx-padding-top: 15;-fx-spacing-top: 10;-fx-background-color:green");
        val footBox = HBox()
        val footInset = Insets(5.0, 0.0, 0.0, 0.0)
        HBox.setMargin(footGridPane, footInset)
        footBox.children.add(footGridPane)
        val borderPane = BorderPane()
        borderPane.center = scrollPane
        borderPane.top = headGridPane
        borderPane.bottom = footBox
        addButton.id = ADDBTN_ID
        addButton.onAction = EventHandler { event: ActionEvent? -> actAddNewAttribute() }
        dialogPane.content = borderPane
    }

    private fun innerInitialize() {
        headGridPane.children.clear()
        val labelList = headerLabelList
        for (cIdx in labelList.indices) {
            headGridPane.add(labelList[cIdx], cIdx, 0)
        }
        gridPane.children.clear()
        orgBodyAttributeNodeList = bodyAttributeNodeList
        for (rIdx in orgBodyAttributeNodeList.indices) {
            val nodeList = orgBodyAttributeNodeList[rIdx]
            for (cIdx in nodeList.indices) {
                gridPane.add(nodeList[cIdx], cIdx, rIdx)
            }
            val wkRowConstraint = RowConstraints()
            wkRowConstraint.minHeight = GRID_MIN_HEIGHT.toDouble()
            gridPane.rowConstraints.add(wkRowConstraint)
        }
        updateFooter()
    }

    private fun setupScreen() {
        val screenSize2d = Screen.getPrimary().visualBounds
        val screenHeight = screenSize2d.height
        val screenWidth = screenSize2d.width
        val window = dialogPane.scene.window
        val stage = window as Stage
        stage.minHeight = 200.0
        stage.minWidth = 300.0
        stage.maxHeight = screenHeight
        stage.maxWidth = screenWidth
    }

    private fun setColumnConstraints() {
        var offset = 0
        for (idx in headerWidthList.indices) {
            val wkWidth = headerWidthList[idx]
            val wkColConstraint = ColumnConstraints()
            wkColConstraint.prefWidth = wkWidth.toDouble()
            gridPane.columnConstraints.add(wkColConstraint)
            headGridPane.columnConstraints.add(wkColConstraint)
            footGridPane.columnConstraints.add(wkColConstraint)
            if (idx != valueColIndex) {
                offset += wkWidth
            }
        }
        scrollPane.hbarPolicy = ScrollBarPolicy.NEVER
        val finalOffset = offset
        val stageSizeListener = ChangeListener { _: ObservableValue<out Number>?, _: Number?, newValue: Number ->
            val newColWidth = newValue.toDouble() - finalOffset - HGAP * (headerWidthList.size - 1)
            gridPane.columnConstraints[valueColIndex].prefWidth = newColWidth
            headGridPane.columnConstraints[valueColIndex].prefWidth = newColWidth
            footGridPane.columnConstraints[valueColIndex].prefWidth = newColWidth
        }
        scrollPane.widthProperty().addListener(stageSizeListener)
    }

    protected fun addAttributeNodeList(newNodeList: List<Node>) {
        val newIdx = orgBodyAttributeNodeList.size + addBodyAttributeNodeList.size
        for (cIdx in newNodeList.indices) {
            gridPane.add(newNodeList[cIdx], cIdx, newIdx)
        }
//        dialogPane.setMaxSize(dialogPane.maxWidth, dialogPane.height + GRID_MIN_HEIGHT.toDouble())
//        gridAnchorPane.maxHeight = gridAnchorPane.maxHeight + GRID_MIN_HEIGHT.toDouble()
//        dialogPane.maxHeight = dialogPane.height + GRID_MIN_HEIGHT.toDouble()
//        val additionalHeight = 30.0
//        dialogPane.resize(dialogPane.width, dialogPane.height  + additionalHeight)
//        gridAnchorPane.resize(gridAnchorPane.width, gridAnchorPane.height + additionalHeight)
        addBodyAttributeNodeList.add(newNodeList)
        updateFooter()
    }

    protected fun updateFooter() {
        footGridPane.children.clear()
        val footerNode = getFooterNodeList()
        for (cIdx in footerNode.indices) {
            footGridPane.add(footerNode[cIdx], cIdx, 0)
        }
    }

    open fun initialize(){
        innerInitialize()
        setColumnConstraints()
        setupScreen()
    }

    companion object {
        const val VALIDATION_MSG_INVALID_VALUE = "Invalid value."
        const val HGAP = 5.0
        const val VGAP = 5.0
        const val FILELD_WIDTH = 300
        const val NAME_COL_WIDTH = 150
        const val DEL_COL_WIDTH = 50
        const val GRID_MIN_HEIGHT = 25
        const val STRFLD_ID_PREFIX = "txtEdit_"
        const val NUMFLD_ID_PREFIX = "numEdit_"
        const val BINFLD_ID_PREFIX = "binEdit_"
        const val VALLBL_ID_PREFIX = "valLabel_"
        const val EDTBTN_ID_PREFIX = "btnEdit_"
        const val ADDBTN_ID = "btnAddAttribute"
        const val DEL_ID_PREFIX = "ckbDel_"
        const val BTN_TITLE = "Edit"
        val NULL_ATTRIBUTE = AttributeValue.builder().nul(true).build()

        /*
        * static function
        */
        // same as DialogPane.createContentLabel
        fun getContentLabel(text: String): Label {
            val label = Label(text)
            label.maxWidth = Double.MAX_VALUE
            label.maxHeight = Double.MAX_VALUE
            label.styleClass.add("content")
            label.isWrapText = false
            label.textOverrun = OverrunStyle.ELLIPSIS
            label.text = text
            return label
        }

        fun getBooleanInput(attrVal: AttributeValue): HBox {
            val hBox = HBox(HGAP)
            val radioButtonTrue = RadioButton()
            val radioButtonFalse = RadioButton()
            radioButtonTrue.text = "TRUE"
            radioButtonFalse.text = "FALSE"
            val toggleGroup = ToggleGroup()
            radioButtonTrue.toggleGroup = toggleGroup
            radioButtonFalse.toggleGroup = toggleGroup
            radioButtonTrue.isSelected = attrVal.bool()
            radioButtonFalse.isSelected = !attrVal.bool()
            hBox.children.addAll(radioButtonTrue, radioButtonFalse)
            return hBox
        }

        val nullViewLabel: Node
            get() = getContentLabel("<null>")

        fun getBooleanValue(hBox: HBox): Boolean {
            val radioButtonTrue = hBox.children[0] as RadioButton
            return radioButtonTrue.isSelected
        }

        fun addUnderlineStyleToNode(node: Node): Node {
            val wkStyle = node.style
            val sbStyle = StringBuilder(wkStyle)
            sbStyle.append("-fx-border-style: none none solid none;")
            sbStyle.append("-fx-border-color: white white lightgray white;")
            sbStyle.append("-fx-border-width: 0 0 1 0;")
            node.style = sbStyle.toString()
            return node
        }
    }

    init {
        this.isResizable = true
        dialogPane.buttonTypes.addAll(ButtonType.OK, ButtonType.CANCEL)
        this.title = dialogTitle
        onCloseRequest = EventHandler { dialogEvent: DialogEvent -> doFinalConfirmation(dialogEvent) }
        dynamoDbRecordOrg = dynamoDbRecord
        setResultConverter { dialogButton: ButtonType ->
            buttonData = dialogButton.buttonData
            if (buttonData == ButtonData.OK_DONE) {
                isValidData = isFinalValidationOk
                return@setResultConverter if (isValidData) editedDynamoDbRecord else emptyAttr
            }
            null
        }
        headGridPane = GridPane()
        gridPane = GridPane()
        gridAnchorPane = AnchorPane()
        scrollPane = ScrollPane()
        footGridPane = GridPane()
        addButton = Button("Add")
        setupComponent()
        // initialize()
    }
}